WIP: workspace file system API

sqs opened the pull request:
I made an attempt at an extension API for reading dirs and files from a repository.

The code is WIP, so don't bother reviewing it for implementation feedback. But API feedback would be helpful.

sqs committed:
refactor: move class URI to shared

felixfbecker reviewed (commented):
I was to design/name this API I would avoid the concept of a "file system" and not use syscall-like names, because that seems like an implementation detail (and doesn't actually match the implementation, since it is backed by GraphQL API and `git show`/`git ls-files`). As a user, I don't care about "reading a file" I care about "getting the content".
I said the same back in the day about "vfs" in LSP :)

felixfbecker commented on src/sourcegraph.d.ts:
```diff
@@ -550,6 +552,88 @@ declare module 'sourcegraph' {
+         *
+         * @todo Document what happens when the root is a subdirectory of a repository.
+         */
+        readonly repository?: {
```
What if the root has multiple repos?

chrismwendt commented on src/sourcegraph.d.ts:
```diff
@@ -550,6 +552,88 @@ declare module 'sourcegraph' {
+             *
+             * This is equivalent to Repository.id in the Sourcegraph GraphQL API schema.
+             */
+            readonly id: string
```
Just wondering, what would the `id` here be used for? Perhaps renaming a repository via the Sourcegraph GraphQL API? Where does an extension get an `id` from?

chrismwendt reviewed (approved):
The API looks good to me! It might be useful to expose a `readDirectoryRecursive` function at some point, but no need for that now.

sqs committed:
feat: parse and expose URIs in extension API

The enhanced sourcegraph.URI class (and its implementation, which is shared by the client and extension) is a full URI parser. Previously it only wrapped a string.

sqs commented:
@felixfbecker:

> I was to design/name this API I would avoid the concept of a "file system" and not use syscall-like names, because that seems like an implementation detail (and doesn't actually match the implementation, since it is backed by GraphQL API and git show/git ls-files)

Are you thinking of changes other than the names? What specific changes do you have in mind?

(This implementation is backed by GraphQL, but other implementations, such as in the editor or when the extension is itself a file system provider (e.g., for vendored or codegenned files) would be more appropriately called "file systems". So I think the name fits, but I am interested to hear your alternatives.)

sqs commented:
> The API looks good to me! It might be useful to expose a readDirectoryRecursive function at some point, but no need for that now.

I added a TODO in the sourcegraph.d.ts file about exposing a way to get all files matching a glob path.

sqs committed:
feat: expose workspace roots and file system API

sqs committed:
fix: wrap extensionHostWorkerMain in try..catch

sqs committed:
chore(deps): upgrade @types/node

chrismwendt commented:
Here's how to move this PR to https://github.com/sourcegraph/sourcegraph/tree/master/packages/sourcegraph-extension-api

```
cd browser-extensions
git format-patch master --stdout > /tmp/patch

cd ../sourcegraph
cat /tmp/patch | git am -3 --directory=packages/sourcegraph-extension-api/
# and fixup merge conflicts
```

felixfbecker commented on src/shared/uri.ts:
```diff
@@ -0,0 +1,142 @@
+ *
+ *   new URL("https://foo/bar").pathname === "/bar"
+ *
+ *   new URL("git://foo/bar").pathname === "//foo/bar" // UNDESIRABLE
```
This is not how WHATWG URL behaves: https://runkit.com/felixfbecker/5be0d65fa5064900123995f4

sqs replied:
Then I think it's not spec compliant (and so soon will change, or will have different behavior in different browsers). Try it in the browser and read the https://url.spec.whatwg.org/#special-scheme doc.
